broadinstitute/scp-ingest-pipeline
# Integrating `ExpressionWriter` into `IngestPipeline` (SCP-4648)

#### BACKGROUND
In order for Image Pipeline to have the necessary pre-rendered expression artifacts available with which to produce static expression scatter plot images, the `ExpressionWriter` class needs to be integrated into the rest of `IngestPipeline` so that it can be launched as a PAPI job.  This means that the CLI needs to be invoked through `ingest_pipeline.py`, and that it can localize/delocalize files from GCP buckets, and have adequate test coverage as a part of normal CI for `scp-ingest-pipeline`.

#### CHANGES
This update changes `ExpressionWriter` from a standalone class to one that can be invoked through `IngestPipeline`.  Basic functionality is unchanged, though file handling is now delegated to the `IngestFiles` class, which can push/pull files directly from GCP buckets.  In addition, logging is now handled by the `monitor.py` module, and MixPanel/Sentry integration is handled through normal codepaths in `IngestPipeline`.   A custom delocalization method was added to handle the expected ~25-30K file export at the end of a successful run.  This will leverage `gsutil` in the host OS to allow for parallel file uploads, as the Python `google.cloud.storage` package does not support this (only serial uploads, and it specifically cannot be parallelized via `multiprocessing`).

#### MANUAL TESTING
1. Pull this branch and initialize the `scp-ingest-pipeline` environment:
```
python3 -m venv env --copies
source env/bin/activate
pip install -r requirements.txt
```
2. Setup MongoDB integration with the following, passing the path to your vault token (we don't need to talk to the database, but ingest will fail to execute without it):
```
source scripts/setup_mongo_dev.sh ~/path/to/github-token-for-vault.txt
```
3. In a separate terminal, get the bucket ID and browser link for any study with the following:
```
study = Study.all.sample # you can use a specific study if you want, it doesn't matter
bucket_id = study.bucket_id
study.google_bucket_url
``` 
4. Copy the URL from above and open it in a browser window, and then copy the following files to the bucket:
  * `tests/data/dense_expression_matrix.txt`
  * `tests/data/cluster_example.txt`
  * `tests/data/mtx/matrix_with_header.mtx`
  * `tests/data/mtx/cluster_mtx_barcodes.tsv`
  * `tests/data/mtx/sampled_genes.tsv`
  * `tests/data/mtx/barcodes.tsv`
5. Using the bucket ID from above, run a job using the dense matrix (the `study_id` and `study_file_id` values aren't used right now, so don't worry about getting valid ones):
```
BUCKET_ID="fc-c36727b0-1663-40fc-ac37-9016a6709829"
python3 ingest_pipeline.py --study-id 5d276a50421aa9117c982845 --study-file-id 5dd5ae25421aa910a723a337 render_expression_arrays \
                             --matrix-file-path gs://$BUCKET_ID/dense_expression_matrix.txt \
                             --matrix-file-type dense \
                             --cluster-file gs://$BUCKET_ID/cluster_example.txt \
                             --cluster-name 'Dense Example' --render-expression-arrays
```
6. You should see output similar to this:
```
distinct_id: 2f30ec50-a04d-4d43-8fd1-b136a2045079
studyAccession: SCPdev
fileName: 5dd5ae25421aa910a723a337
fileType: input_validation_bypassed
fileSize: 1
trigger: dev-mode
logger: ingest-pipeline
appId: single-cell-portal
(env) bistline@wm306-020 ingest % If you experience problems with multiprocessing on MacOS, they might be related to https://bugs.python.org/issue33725. You can disable multiprocessing by editing your .boto config or by adding the following flag to your command: `-o "GSUtil:parallel_process_count=1"`. Note that multithreading is still available even if you disable multiprocessing.

Copying file://Dense_Example/Setd6.json.gz [Content-Type=application/json]...
Copying file://Dense_Example/Gm12000.json.gz [Content-Type=application/json]... 
Copying file://Dense_Example/Sergef.json.gz [Content-Type=application/json]...  
Copying file://Dense_Example/Itm2a.json.gz [Content-Type=application/json]...   
Copying file://Dense_Example/1700061E18Rik.json.gz [Content-Type=application/json]...
Copying file://Dense_Example/Fam71e2.json.gz [Content-Type=application/json]... 
Copying file://Dense_Example/Chil5.json.gz [Content-Type=application/json]...   
Copying file://Dense_Example/Gstm2.json.gz [Content-Type=application/json]...   
Copying file://Dense_Example/Fam109a.json.gz [Content-Type=application/json]... 
Copying file://Dense_Example/Gm14444.json.gz [Content-Type=application/json]... 
Copying file://Dense_Example/Vps28.json.gz [Content-Type=application/json]...
Copying file://Dense_Example/Hebp2.json.gz [Content-Type=application/json]...   
Copying file://Dense_Example/Psma4.json.gz [Content-Type=application/json]...   
Copying file://Dense_Example/Mks1.json.gz [Content-Type=application/json]...    
Copying file://Dense_Example/Ssu72.json.gz [Content-Type=application/json]...   
Copying file://Dense_Example/Eif2b2.json.gz [Content-Type=application/json]...  
Copying file://Dense_Example/Olfr1018.json.gz [Content-Type=application/json]...
Copying file://Dense_Example/Spn-ps.json.gz [Content-Type=application/json]...  
Copying file://Dense_Example/Dhx9.json.gz [Content-Type=application/json]...    
/ [19/19 files][  1.1 KiB/  1.1 KiB] 100% Done                                  
Operation completed over 19 objects/1.1 KiB.                                     
```
7. In the browser window with the bucket, refresh the page and navigate to `_scp_internal/cache/expression_scatter/data/Dense_Example`
8. Confirm there are two files here: `Itm2a.json.gz` and `Sergef.json.gz`
9. Now run the sparse example:
```
BUCKET_ID="fc-c36727b0-1663-40fc-ac37-9016a6709829"
python3 ingest_pipeline.py --study-id 5d276a50421aa9117c982845 --study-file-id 5dd5ae25421aa910a723a337 \
                             render_expression_arrays --matrix-file-path gs://$BUCKET_ID/matrix_with_header.mtx \
                             --matrix-file-type mtx \
                             --gene-file gs://$BUCKET_ID/sampled_genes.tsv \
                             --barcode-file gs://$BUCKET_ID/barcodes.tsv \
                             --cluster-file gs://$BUCKET_ID/cluster_mtx_barcodes.tsv \
                             --cluster-name 'Sparse Example' --render-expression-arrays
```
10. Confirm similar output to above, and back in the bucket view, validate there are 7 new files for the following genes: `C1orf159`, `RP11-345P4.9`, `GABRD`, `THAP3`, `DNAJC11`, `OXCT2`, `HOMER2`
